shrinkwrap: exclude dependencies of devDependencies and optionalDependencies

terinjokes opened the pull request:
As the tree is now flattened, dependencies of devDependencies weren't
automatically excluded simply by not recursing into a devDependency's
children.

Instead, now explicitly walk "up" a dependency's tree to the root to
determine if it's a devDependency or solely a dependency of
a devDependency.

Resolve #9727

iarna commented on lib/install/is-dev.js:
```diff
@@ -2,6 +2,26 @@
   return node.package._requiredBy.some(function (req) { return req === '#DEV:/' })
 }
-exports.isOnlyDev = function (node) {
+var isOnlyDev = exports.isOnlyDev = function (node) {
```
We can stop exporting `isOnlyDev` as this is the only use, or even inline it directly into `isTreeDev`.

terinjokes replied:
@iarna I'm happy to do either, is there a preference?

terinjokes replied:
@iarna sorry for the double comment :crying_cat_face: The other export, `isDev` is only used by is-extraneous.js, but after a quick glance, it doesn't look necessary. Mind taking a look as well?

iarna replied:
@terinjokes The `is-extraneous` use sure looks needed to me. Without it, all dev-only deps would be listed as extraneous.

iarna replied:
(And my inclination would be to just inline `isOnlyDev`, to answer your first question.)

terinjokes replied:
Sorry, I think when I glanced I reversed the booleans :(

fiznool commented:
Any ideas when this might get looked at? The underlying issue is preventing me from using npm3, since using shrinkwrap is erroneously including devDependencies.

terinjokes committed:
shrinkwrap: exclude dependencies of devDependencies

As the tree is now flattened, dependencies of devDependencies weren't
automatically excluded simply by not recursing into a devDependency's
children.

Instead, now explicitly walk "up" a dependency's tree to the root to
determine if it's a devDependency or solely a dependency of
a devDependency.

Resolve #9727

terinjokes committed:
shrinkwrap: inline `isOnlyDev` into `isTreeDev`

terinjokes commented:
(rebased on top of master)

terinjokes commented:
I've added logic to handle optional dependencies (right now just by excluding them, can modify the what later, perhaps as part of #10073)

Unfortunately, I'm leaving for lunch, so you'll have to wait a few more minutes. :)

terinjokes committed:
shrinkwrap: understand and exclude optionalDependencies

Optional dependencies tend to be dependencies that are nice to include,
but aren't required for the operation of a project. Commonly used for
dependencies that are needed only on one platform.

If a shrinkwrap is created on a platform where an optional dependency is
installed, the optional dependencies are then included in the
shrinkwrap. These dependencies may fail to install on other platforms,
causing the whole install to be in an indeterminate state.

This patch adds logic to detect optional dependencies while turning the
dependency tree into a shrinkwrap file as well as excludes them from the
shrinkwrap.

iarna commented:
I don't really like the `#OPT` thing as logic. I'm moving the installer away from using the `package.json` `_requiredBy` list entirely (in favor of requires & requiredBy on the tree object).

terinjokes commented:
@iarna Is any of that setup so that I can refactor this to use that?

iarna commented:
Yeah, it would look something like:
```
var isOpt = exports.isOpt = function (node) {
  return node.requiredBy.some(function (req) {
    return req.package.optionalDependencies[node.package.name]
  })
}
```

terinjokes commented:
I'll work on moving both (but most importantly, devDeps) over to that.
Thanks for the little bit of insight.

-- 
Terin Stock

On Wed, Jan 6, 2016 at 2:17 PM, Rebecca Turner <notifications@github.com>
wrote:

> Yeah, it would look something like:
>
> var isOpt = exports.isOpt = function (node) {
>   return node.requiredBy.some(function (req) {
>     return req.package.optionalDependencies[node.package.name]
>   })
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/npm/npm/pull/10265#issuecomment-169479921>.
>

johnman commented:
Is there a rough estimate for when this will be merged? For modules that have lots of dev dependencies (e.g. grunt, plugins etc) but little direct dependencies it can make the shrinkwrapped file much bigger than it needs to be. Thanks!

terinjokes commented:
@johnman This is something I'm going to get done, one way or another, sooner or later. However, I don't have a timeline for redoing this change. I've got a lot of things going on in my life—and I'm trying to be better at finding time to relax too!

I realize that's probably not the answer you were hoping for. I'm happy to bump this up in my priority list if we can come to an agreement. Otherwise, I'll get to it.

johnman commented:
Hi, it would be useful and I noticed that there is another pull request with a similar theme but I think the only thing you should be bumping up your priority list is sorting out things going on in your life and finding time to relax :) I appreciate the quick response but you do it when you have the time (I've got an idea for post processing the shrink wrap file via a grunt task to clean it up in the meantime).

terinjokes commented:
In light of, uhm, recent events, I'm going to attempt to resolve the issues with this branch tonight.

htulipe commented:
Any news on this fix @terinjokes ?

terinjokes commented:
@htulipe still in progress, unfortunately. I'm really sorry 😢

htulipe commented:
Hey no need to apologize @terinjokes, this is an opensource project ;)

iarna commented:
I'm (probably) landing #10073 later today in our release staging branch and thanks to that, I think like 90% of this will be done? Optional deps (including transitives) will be flagged so it'll be easy to add code to exclude them if need be.

terinjokes commented:
@iarna Yeah, the two pull requests ended up being pretty similar. I'm going to go ahead and close this, but let me know if I should look at adding any of the optionalDependencies logic.
